internal/common/tree.go

```diff
@@ -0,0 +1,201 @@
+package common
+
+import (
+	"slices"
+
+	"sigs.k8s.io/external-dns/endpoint"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+)
+
+type DNSTreeNode struct {
+	Name     string
+	Children []*DNSTreeNode
+	DataSets []DNSTreeNodeData
+}
+
+type DNSTreeNodeData struct {
+	RecordType       string
+	SetIdentifier    string
+	RecordTTL        endpoint.TTL
+	Labels           endpoint.Labels
+	ProviderSpecific endpoint.ProviderSpecific
+	Targets          []string
+}
+
+// RemoveNode removes a node from a tree.
+// If the node was the only child of the parent node,
+// the parent will be removed as well unless the parent is a root node
+func (n *DNSTreeNode) RemoveNode(deleteNode *DNSTreeNode) {
+	// store indexes of dead branches
+	var deadBranches []int
+	var deadDataSets []string
+
+	if deleteNode == nil {
+		return
+	}
+
+	for i, node := range n.Children {
+		if node.Name == deleteNode.Name {
+			n.Children = append(n.Children[:i], n.Children[i+1:]...)
+
+			// we removed child, but need to clean up data sets
+			for j, dataset := range n.DataSets {
+				index := slices.Index(dataset.Targets, deleteNode.Name)
+				if index >= 0 {
+					n.DataSets[j].Targets = append(dataset.Targets[:index], dataset.Targets[index+1:]...)
+				}
+			}
+			return
+		}
+
+		// no children matched, try on each child
+		node.RemoveNode(deleteNode)
+
+		// the removed node was the only child - we have a dead branch
+		// children are nil on leaf node
+		// not checking for it will nuke the whole tree as leafs will be considered dead branches
+		if node.Children != nil && len(node.Children) == 0 {
+			// store the index. indexes are in ascending order
+			deadBranches = append(deadBranches, i)
+
+			// we can't rely on indexes for data sets, so store node name
+			deadDataSets = append(deadDataSets, node.Name)
+		}
+	}
+
+	// prune dead branches separately from the main for loop.
+	// doing it inside will shift indexes the for loop is iterating through
+	for count, deadBranchIndex := range deadBranches {
+		// after the first removal, all subsequent deadBranchIndexes will de one to high,
+		// but since we have them ascending, we can use count as modifier
+		n.Children = append(n.Children[:deadBranchIndex-count], n.Children[deadBranchIndex-count+1:]...)
+	}
+
+	var healthyDataSets []DNSTreeNodeData
+	// clean up data nodes from dead branches
+	for _, dataSet := range n.DataSets {
+
+		// we are dealing with CNAMES only here.
+		// the A record is already removed from datasets
+		if !slices.Contains(deadDataSets, dataSet.Targets[0]) {
```

Comment:
why targets[0]?

Reply:
ok because it is a CNAME record? I am not following how we know it is CNAME though

Reply:
if it will be A record we will return much earlier. I could add a check here but it is redundant